Support for BlackWidow V4 Pro

Jacajack opened the pull request:
Current status:
- [x] Reading serial number
- [x] Brightness control
- [x] Basic matrix effects (had to use `WAVE_DIRS` to reverse wave effect direction)
- [x] Custom matrix effects
- [x] Gamemode LED control 
- [x] Wheel matrix effect
- [x] Macro LED control (turns out it works fine when the daemon is active)
- [x] Command dial bindings (can be handled with an external script, see [this comment](https://github.com/openrazer/openrazer/pull/2054#issuecomment-1603412688))
- [x] <kbd>M1</kbd>-<kbd>M5</kbd> macro key support
- [x] <kbd>M6</kbd>-<kbd>M8</kbd> macro key support (side buttons)
- [x] Basic wheel effect support
- [x] Volume acts as scroll  
- [x] Entering/exiting gamemode is broken _sometimes_
- [x] Onboard macros seem to be broken (mapped macro keys to F13-F20, F24)
- [x] Update tests etc.

Currently all M1-M5 bindings set in Synapse work on Linux as well, however all non-standard bindings for the command dial are lost. It would be great if we could get the driver to enable them back on Linux. Not sure if this should become a part of this PR, though.

I'm currently investigating why the macro LED doesn't work. I've tweaked `razer_attr_write_macro_led_state()` so it sends the same commands as Synapse, but with no luck so far.  Synapse does it like that (`NOSTORE` and transaction ID `0x1f` seem to be a bit unusual):

<img src=https://user-images.githubusercontent.com/6173262/234650093-2ebd4b35-61c7-4994-b6b4-0191a6073ab5.png width=50% />





This should close #2044.

z3ntu commented on driver/razerkbd_driver.h:
```diff
@@ -94,6 +94,7 @@
 #define USB_DEVICE_ID_RAZER_DEATHSTALKER_V2_PRO_WIRED 0x0292
 #define USB_DEVICE_ID_RAZER_DEATHSTALKER_V2 0x0295
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_TK 0x0A24
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V4_PRO 0x028D
```
please keep this list sorted by USB PID, so put this just below USB_DEVICE_ID_RAZER_BLADE_14_2022

Jacajack replied:
Fixed in 1a9965bc346bb01841ffa1216f09b26d3ed00867

Jacajack committed:
Add wheel effect support in daemon and pylib

Driver support will be added in an upcoming commit.

z3ntu commented on driver/razerkbd_driver.c:
```diff
@@ -1516,6 +1525,32 @@ static ssize_t razer_attr_write_matrix_effect_wave(struct device *dev, struct de
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V4_PRO:
+        request = razer_chroma_extended_matrix_effect_wheel(VARSTORE, BACKLIGHT_LED, direction);
+        request.transaction_id.id = 0x1F;
+        break;
```
please add a block like this
```c
    default:
        printk(KERN_WARNING "razerkbd: matrix_effect_wheel not supported for this model\n");
        return -EINVAL;
```

Jacajack committed:
razerkbd: Don't ask for response for BlackWidow V4 Pro custom frame

Do not expect responses for `matrix_custom_frame` and
`matrix_effect_custom` commands. This is more similar to what Synapse
does on Windows and apparently eliminates matrix glitches when
continuously updating a custom effect.

See: https://github.com/openrazer/openrazer/pull/2054#issuecomment-1615283977

z3ntu commented on daemon/openrazer_daemon/keyboard.py:
```diff
@@ -134,7 +134,9 @@
-    183: 'M1', 184: 'M2', 185: 'M3', 186: 'M4', 187: 'M5', 188: 'MACROMODE', 189: 'GAMEMODE', 190: 'BRIGHTNESSDOWN', 194: 'BRIGHTNESSUP'
+    183: 'M1', 184: 'M2', 185: 'M3', 186: 'M4', 187: 'M5', 188: 'M6',
+    189: 'M7', 190: 'M8', 191: 'M9', 192: 'M10', 193: 'M11', 194: 'M12',
+    0x2ad: 'MACROMODE', 0x2ac: 'GAMEMODE', 0x2ab: 'BRIGHTNESSDOWN', 0x2aa: 'BRIGHTNESSUP'
```
Any reason these keycodes are in hex while the rest is in dec?

Jacajack replied:
I thought it would be less confusing if the codes here matched with definitions in [linux/input-event-codes.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h). I can change that to dec if you like.

z3ntu commented on daemon/openrazer_daemon/misc/key_event_management.py:
```diff
@@ -356,8 +356,8 @@ def key_action(self, event_time, key_id, key_press='press'):
-            # 190 (KEY_F20) and 194 (KEY_F24) are defined in razerkbd_driver.c
-            if key_id in (190, 194):
+            # Brightness keys are defined in keyboard.py and razerkbd_driver.c
+            if key_id in (0x2ab, 0x2aa):
```
`0x2aa, 0x2ab` would make more sense, otherwise it's descending :wink:

z3ntu commented on driver/razerkbd_driver.c:
```diff
@@ -11,6 +11,7 @@
 #include <linux/hid.h>
 #include <linux/dmi.h>
 
+#include "linux/input-event-codes.h"
```
Shouldn't this be `#include <linux/input-event-codes.h>` ?

Jacajack replied:
It definitely should. `clangd` must have added that when I wasn't looking. I will fix that

z3ntu commented on driver/razerkbd_driver.c:
```diff
@@ -27,25 +28,39 @@ MODULE_DESCRIPTION(DRIVER_DESC);
+#define RAZER_MACRO_KEY KEY_MACRO30 // TODO maybe KEY_MACRO_RECORD_START?
+#define RAZER_GAME_KEY KEY_MACRO29 // TODO maybe KEY_GAMES?
+#define RAZER_BRIGHTNESS_DOWN KEY_MACRO28
+#define RAZER_BRIGHTNESS_UP KEY_MACRO27
```
Ah I guess you're going bottom up basically, if we need new keys we take 26 etc. Still not sure if I'd not rather have down = 27, up = 28 here

Jacajack replied:
Yes, I've decided to map Chroma keys to higher macro keys and leave ascending numeration for the actual macro keys (just in case). I will swap `RAZER_BRIGHTNESS_DOWN` and `RAZER_BRIGHTNESS_UP` then.

z3ntu commented on driver/razerkbd_driver.c:
```diff
@@ -27,25 +28,39 @@ MODULE_DESCRIPTION(DRIVER_DESC);
+#define RAZER_BRIGHTNESS_UP KEY_MACRO27
+#define RAZER_FN 195
+
+#define KEY_FLAG_BLOCK 0b00000001
```
I've removed this define and RAZER_FN in git master now since they're not used at all. Will rebase later (or you can do it too)

z3ntu commented on driver/razerkbd_driver.c:
```diff
@@ -3162,13 +3177,7 @@ static int razer_event(struct hid_device *hdev, struct hid_field *field, struct
-        }
-
-        if(do_translate) {
+        if (test_bit(usage->code, device->pressed_fn) || device->fn_on) {
```
Nice suggestion, pushed this bit to git master now already.

z3ntu commented on pylib/openrazer/client/fx.py:
```diff
@@ -157,6 +157,27 @@ def wave(self, direction: int) -> bool:
+        :raises ValueError: If direction is invalid
+        """
+        if direction not in (c.WAVE_LEFT, c.WAVE_RIGHT):
+            raise ValueError("Direction must be WAVE_RIGHT (0x01) or WAVE_LEFT (0x02)")
```
Not sure if it's a good idea to use WAVE_DIRS now and cement this into the basically public API. Could you add WHEEL_LEFT / WHEEL_RIGHT please?

Jacajack replied:
Sure, I will add these

z3ntu commented on driver/razerkbd_driver.c:
```diff
@@ -290,24 +291,54 @@ static int razer_get_report(struct usb_device *usb_dev, struct razer_report *req
+        *report_index = 0x01;
+        *response_index = 0x01;
+        *wait_min = RAZER_BLACKWIDOW_CHROMA_WAIT_MIN_US;
+        *wait_max = RAZER_BLACKWIDOW_CHROMA_WAIT_MAX_US;
```
Maybe you've had this diff in a separate commit and I lost it during the rebase/squash, but would be nice if the no-change refactor would be a separate commit (without razer_send_payload_no_response then)

Jacajack replied:
The BlackWidow V4 workaround was commited separately here https://github.com/Jacajack/openrazer/commit/dbb069078c5a41eb617388da24a5eb93478f97f7. If you like, I can re-squash the commits differently.

Jacajack replied:
I managed to "unsquash" that commit.

Jacajack committed:
Add support for Razer BlackWidow V4

Co-authored-by: Luca Weiss <luca@z3ntu.xyz>

Jacajack committed:
razerkbd: refactor razer_get_report function

Prepare to have a slighly different implementation for razer_get_report.

z3ntu reviewed (approved):
Reworked a few things but LGTM now. Thanks a lot for the PR!
